diego-urgell/BinSeg
# Cost function for both remaining normal distribution changes

# Description 

In this branch I will implement the cost functions for the following distributions: 
 
- Normal with change in variance
- Normal with change in mean and variance 

Currently, mean and variance is implemented. However, there have been complications. 

# Tests 

I added three tests to the `test-BinarySegmentationTests.R` file in order to try three different types of data with the NormMeanVar distribution: 

1. Change in mean with constant variance 
2. Change in variance with constant mean 
3. Change in mean and variance 

The first test passes almost always, while the second one pases around 90% of times. However, the third test fails very frequently, yielding a change point of 18. I have been debugging this a couple of days, but I haven't managed to solve it. 

Anyway, using `lldb` with a Python script, I found something pretty interesting. The data I set was a vector of length 20, using the parameters from test 3. The value of the `i` counter begins on 1 and ends in 18 on the first execution of this function. 

The script was designed to print the `currSplitCost` and the `this  -> mid` variables on files `Segments.cpp` line 48, which is part of the `optimalPartition` function. The following screenshot shows the values: 

<img width="1703" alt="Screen Shot 2021-06-29 at 23 54 52" src="https://user-images.githubusercontent.com/45611081/124035413-1db1b380-d9c2-11eb-8112-194acf54d567.png">

As you can note, on the first execution of the function (before curr_mid returns to 1), the last value of curr_mid is 9 (which is correct), and the costs indeed show that this is the optimal partition. However, when this process is finished and I check the returned answer from the R console, I get the following (ignore the first lines, since they were of unused executions of optimalPartition function): 

<img width="1703" alt="Screen Shot 2021-06-29 at 23 54 55" src="https://user-images.githubusercontent.com/45611081/124035642-71bc9800-d9c2-11eb-927d-e4fbc78ee813.png">

Clearly there is something going on that changes the cpts value to 18 despite optimalPartition finding the correct one. However, I haven't found where this happens exactly. I will continue working on this because is probably a mistake we would face on the rest of the distributions. 

For now, I just wanted to let you know what was going on. @tdhock @rkillick , do you have any idea where the error could be? 

Note: I explicitly created the BS and NormMeanVar classes instead of using the macro because it is impossible to test with lldb otherwise. When everything is working I will switch them both to the Macro.
